### PR TITLE
feat: add bff using FastAPI

### PR DESCRIPTION
**Pull request by makram93**
Added initial commit.

TODO:
- [x] Correct the request/response format
- [x] Match the endpoints url
- [ ] Docker containerised - TODO: In a separate PR
- [x] Implement the search - and two use-cases - Image, Texts
- [x] Add unit test cases for the endpoints
- [x] Add integration test

Layout:
image

### CODE REVIEW

**Comment by florian-hoenicke** [comment]
I would recommend combining `/search/{query}` with `/search/`
I think it should be one endpoint with different parameters.
Otherwise, it might be confusing to the user.
Instead of doing the binary file upload, you could just have one json endpoint for search which gets base 64 encoded images/ text

**Review by florian-hoenicke — CHANGES_REQUESTED** [review]
I would recommend combining /search/{query} with /search/
I think it should be one endpoint with different parameters.
Otherwise, it might be confusing to the user.
Instead of doing the binary file upload, you could just have one json endpoint for search which gets base 64 encoded images/ text

File: now/bff/v1/routers/image.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    limit: int = 10,
+):
+    """
+    Retrieve matching images for a given image uri as query
+    """
+    # TODO: Uri or FileUploader?
+    contents = image_file.file.read()
+    query_doc = Document(blob=contents)
+    query_doc.convert_blob_to_image_tensor(224, 224)
+    c = Client(host=host, port=31080)
```
Comment (bwanglzu) [inline now/bff/v1/routers/image.py:63]: what will happen if you have 100 users calling the endpoint? everytime you initialize a new Client instance?

**Review by bwanglzu — COMMENTED** [review]
left one comment, btw, do you need to handle user authentification?

**Comment by makram93** [comment]
> left one comment, btw, do you need to handle user authentification?

No, we do not need any authentication for now

File: requirements-dev.txt [hunk 2]
```diff
@@ -1,2 +1,3 @@
 pytest
 pytest-mock==3.7.0
+pytest==0.75.2
```
Comment (joschkabraun) [inline requirements-dev.txt:3]: shall we remove the first line then?

File: now/bff/v1/routers/image.py (hunk 1 above)
Comment (makram93) [reply now/bff/v1/routers/image.py:63]: We do not have a similar scenario as Finetuner API. As in our case, Frontend and BFF will be deployed for each user running jina now and then they can call the endpoints for their deployed Flow (either local or GKE).

File: now/bff/v1/routers/image.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    host: str = 'localhost',
+    limit: int = 10,
+):
+    """
+    Retrieve matching images for a given image uri as query
+    """
+    # TODO: Uri or FileUploader?
+    contents = image_file.file.read()
+    query_doc = Document(blob=contents)
+    query_doc.convert_blob_to_image_tensor(224, 224)
```
[the remark this answers is not in the document]
Comment (makram93) [reply now/bff/v1/routers/image.py]: Thanks for pointing this out. I have removed it and unified the endpoints and now it simply accepts base64 encoded image/text. So, whatever resizing is needed should be done at the client/Flow and not be handled in the endpoint logic

File: requirements-dev.txt (hunk 2 above)
Comment (makram93) [reply requirements-dev.txt:3]: Good observation

File: now/bff/v1/routers/image.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    response_model=JinaResponseModel,
+    summary='Add more data to the indexer',
+)
+def index(data: List[str], host: str):
+    """
+    Append the image data to the indexer
+    """
+    index_docs = DocumentArray()
+    for text in data:
+        index_docs.append(Document(text=text))
```
[the remark this answers is not in the document]
Comment (makram93) [reply now/bff/v1/routers/image.py:30]: Which part? It is intentional `text=text`

**Comment by florian-hoenicke** [comment]
I ran it locally and checked the response of the bff.
It returns a DocumentArray at the moment.
However the ResponseModel should also be adjusted to the specific usecase:
For Instance, when using text as output modality, the user would expect a responsemodel like:
```
[
  {
    id: 3457483,
    text: first text,
    tags: {author: name, url:https://...}
  },
  {
    text:...
]
```
image

**Pull request merged**